src/repository/x509.rs

```diff
@@ -103,94 +91,110 @@ impl Name {
+                            if sn {
+                                return Err(cons.content_err(
+                                    "invalid name"
+                                ))
                             }
-                            Ok(())
-                        })? { }
+                            let _ = PrintableString::take_from(cons)?;
+                            sn = true;
+                        }
                         Ok(())
-                    })? {}
-                    Ok(())
-                })?;
-                // Common name is MUST.
-                if cn {
+                    })? { }
                     Ok(())
-                }
-                else {
-                    Err(decode::Error::Malformed)
-                }
+                })? {}
+                Ok(())
+            })?;
+            // Common name is MUST.
+            if cn {
+                Ok(())
+            }
+            else {
+                Err(cons.content_err("invalid name"))
+            }
+        }
+
+        if strict {
+            self.0.clone().decode(
+                inspect_strict
+            ).map_err(|_| {
+                InspectionError::new("invalid name")
             })?
         }
         Ok(())
     }
 
     /// Validate the name to conform with BGPSec router certificates.
-    pub fn validate_router(
+    pub fn inspect_router(
         &self, strict: bool
-    ) -> Result<(), ValidationError> {
-        if strict {
-            self.0.clone().decode(|cons| {
-                let mut cn = false;
-                let mut sn = false;
-                cons.take_sequence(|cons| {
-                    while let Some(()) = cons.take_opt_set(|cons| {
-                        while let Some(()) = cons.take_opt_sequence(|cons| {
-                            let id = Oid::take_from(cons)?;
-                            if id == oid::AT_COMMON_NAME {
-                                if cn {
-                                    xerr!(
-                                        return Err(decode::Error::Malformed)
-                                    )
-                                }
-                                Self::skip_router_string(cons)?;
-                                cn = true;
+    ) -> Result<(), InspectionError> {
+        fn inspect_strict<S: decode::Source>(
+            cons: &mut decode::Constructed<S>
+        ) -> Result<(), DecodeError<S::Error>> {
+            let mut cn = false;
+            let mut sn = false;
+            cons.take_sequence(|cons| {
+                while let Some(()) = cons.take_opt_set(|cons| {
+                    while let Some(()) = cons.take_opt_sequence(|cons| {
+                        let id = Oid::take_from(cons)?;
+                        if id == oid::AT_COMMON_NAME {
+                            if cn {
+                                return Err(cons.content_err(
+                                    "invalid router subject"
```

Comment:
Subject? But this is not the SN, this is the CN, is subject the right term here?

Reply:
None of these actually mattered, because the error was mapped away in the outer function. However, I changed this now and the errors are carried through and the strings have been made more precise as suggested.